src/data/finsupp/ne_locus.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import data.finsupp.basic
+
+/-!
+#  `ne_locus` for finitely supported functions
+
+Let `α N` be two Types, and assume that `N` has a `0` and let `f g : α →₀ N` be finitely supported
+functions.
+
+## Main definition
+
+* `finsupp.ne_locus f g : finset α`, the finite subset of `α` where `f` and `g` differ.
+
+In the case in which `N` is an additive group, `finsupp.ne_locus f g` coincides with
+`finsupp.support (f - g)`.
+-/
+
+variables {α M N P : Type*}
+
+namespace finsupp
+variables [decidable_eq α] [decidable_eq N] [decidable_eq M] [has_zero M]
+  [decidable_eq P] [has_zero P]
+
+section ne_locus
+
+section N_has_zero
+variables [has_zero N] {f g : α →₀ N}
+
+/--  Given two finitely supported functions `f g : α →₀ N`, `finsupp.ne_locus f g` is the `finset`
+where `f` and `g` differ. This generalizes `(f - g).support` to situations without subtraction. -/
+def ne_locus (f g : α →₀ N) : finset α :=
+(f.support ∪ g.support).filter (λ x, f x ≠ g x)
+
+@[simp]
+lemma mem_ne_locus {a : α} : a ∈ f.ne_locus g ↔ f a ≠ g a :=
+by simpa only [ne_locus, finset.mem_filter, finset.mem_union, mem_support_iff,
+    and_iff_right_iff_imp] using ne.ne_or_ne _
+
+lemma coe_ne_locus : ↑(f.ne_locus g) = {x | f x ≠ g x} :=
+by { ext, exact mem_ne_locus }
+
+@[simp] lemma ne_locus_eq_empty : f.ne_locus g = ∅ ↔ f = g :=
+begin
+  refine ⟨λ h, _, λ h, h ▸ by simp [ne_locus]⟩,
+  ext a,
+  exact not_not.mp (mem_ne_locus.not.mp (finset.eq_empty_iff_forall_not_mem.mp h a)),
+end
+
+@[simp] lemma nonempty_ne_locus_iff : (f.ne_locus g).nonempty ↔ f ≠ g :=
+finset.nonempty_iff_ne_empty.trans ne_locus_eq_empty.not
+
+variables (f g)
+
+lemma ne_locus_comm (f g : α →₀ N) : f.ne_locus g = g.ne_locus f :=
+by simp_rw [ne_locus, finset.union_comm, ne_comm]
+
+@[simp]
+lemma ne_locus_zero_right : f.ne_locus 0 = f.support :=
+by { ext, rw [mem_ne_locus, mem_support_iff, coe_zero, pi.zero_apply] }
+
+@[simp]
+lemma ne_locus_zero_left : (0 : α →₀ N).ne_locus f = f.support :=
+(ne_locus_comm _ _).trans (ne_locus_zero_right _)
+
+lemma subset_map_range_ne_locus {F : N → M} (F0 : F 0 = 0) :
+  (f.map_range F F0).ne_locus (g.map_range F F0) ⊆ f.ne_locus g :=
+begin
+  refine λ x, _,
+  simp only [mem_ne_locus, map_range_apply, not_imp_not],
+  exact congr_arg _,
+end
+
+lemma map_range_ne_locus_eq {F : M → N → P} (F0 : F 0 0 = 0)
```

Comment:
Whoops, I meant to name this
```suggestion
lemma zip_with_ne_locus_eq_left {F : M → N → P} (F0 : F 0 0 = 0)
```

Reply:
Whether you decide to call this one `left` or `right`, we should have both versions.

Reply:
I went with `right` for the one that you had written and `left` for the other.  My rationale was to match the chirality of the `add_cancel`, rather than the `injective` one.  I am always confused by these conventions.